File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.cc
#include <iostream>
#include <vector>
#include <deque>
#include <algorithm>

using namespace std;

void dfs(deque<deque<int>> &nodeValues, vector<bool> &visited, deque<int> graphTool, int V)
{
	graphTool.push_front(V);

	while (!graphTool.empty())
	{
		V = graphTool.front();
		graphTool.pop_front();

		if (visited[V])
		{
			continue;
		}

		cout << V << " ";
		visited[V] = true;

		for (int index : nodeValues[V])
		{
			graphTool.push_front(index);
		}
	}
}

void bfs(deque<deque<int>>& nodeValues, vector<bool>& visited, deque<int> graphTool, int V)
{
	graphTool.push_back(V);

	while (!graphTool.empty())
	{
		V = graphTool.front();
		graphTool.pop_front();

		if (visited[V])
		{
			continue;
		}

		cout << V << " ";
		visited[V] = true;

		for (int index : nodeValues[V])
		{
			graphTool.push_back(index);
		}
	}
}

int main()
{
	ios_base::sync_with_stdio(false);
	cin.tie(nullptr);

	int N, M, V;

	cin >> N >> M >> V;

	deque<deque<int>> nodeValues(N + 2);
	vector<bool> visited(N + 2);

	deque<int> graphTool;	// DFS, BFS가 그래프 이론으로 분류된다고 함

	for (int i = 0; i < M; i++)
	{
		int firstNode, secondNode;

		cin >> firstNode >> secondNode;

		nodeValues[firstNode].push_back(secondNode);
		nodeValues[secondNode].push_back(firstNode);
	}

	for (int i = 0; i <= N; i++)
	{
		sort(nodeValues[i].begin(), nodeValues[i].end(), greater<>());
	}

	dfs(nodeValues, visited, graphTool, V);

	cout << endl;

	for (int i = 0; i <= N; i++)
	{
		visited[i] = false;
	}

	for (int i = 0; i <= N; i++)
	{
		reverse(nodeValues[i].begin(), nodeValues[i].end());
	}

	bfs(nodeValues, visited, graphTool, V);

	return 0;
}

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.py
import sys
from collections import deque

# def dfs(graph, v, visited):
#     visited[v] = True
#     print(v, end=' ')

#     for i in graph[v]:
# if not visited[i]:
#             dfs(graph, i, visited)


def dfs(graph, v, visited):
    stack = deque([v])

    while stack:
        current = stack.pop()

        if not visited[current]:
            visited[current] = True
            print(current, end=" ")

            for i in reversed(graph[current]):
                if not visited[i]:
                    stack.append(i)


def bfs(graph, start, visited):
    queue = deque([start])
    visited[start] = True

    while queue:
        v = queue.popleft()
        print(v, end=" ")

        for i in graph[v]:
            if not visited[i]:
                queue.append(i)
                visited[i] = True


# N: 정점의 개수, M: 간선의 개수, V: 시작 정점 번호
N, M, V = map(int, sys.stdin.readline().split())

graph = [[] for _ in range(N + 1)]

for _ in range(M):
    x, y = map(int, sys.stdin.readline().split())
    graph[x].append(y)
    graph[y].append(x)

for index in range(N + 1):
    graph[index].sort()  # 정렬 이유: 정점 번호가 작은 것을 먼저 방문해야 하기 때문

visited = [False] * (N + 1)
dfs(graph, V, visited)

print()

visited = [False] * (N + 1)
bfs(graph, V, visited)